src/Raven.Server/Documents/PeriodicBackup/Restore/ShardedRestoreOrchestrationTask.cs

```diff
@@ -0,0 +1,177 @@
+        }
+
+        protected override async Task Restore()
+        {
+            var dbSearchResult = ServerStore.DatabasesLandlord.TryGetOrCreateDatabase(DatabaseName);
+            var shardedDbContext = dbSearchResult.DatabaseContext;
+
+            _result = await RestoreOnAllShards(Progress, RestoreConfiguration, shardedDbContext);
+        }
+
+        protected override void OnAfterRestore()
+        {
+            DisableOngoingTasksIfNeeded(RestoreSettings.DatabaseRecord);
+
+            Progress.Invoke(Result.Progress);
+
+            // after the db for restore is done, we can safely set the db state to normal and write the DatabaseRecord
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext serverContext))
+            using (serverContext.OpenReadTransaction())
+            {
+                var databaseRecord = ServerStore.Cluster.ReadDatabase(serverContext, DatabaseName);
+                databaseRecord.DatabaseState = DatabaseStateStatus.Normal;
+
+                RestoreSettings.DatabaseRecord = databaseRecord;
+            }
+        }
+
+        protected override IOperationResult OperationResult() => _result;
+
+        private void InitializeShardingConfiguration()
+        {
+            var databaseRecord = RestoreSettings.DatabaseRecord;
+
+            databaseRecord.Sharding = new ShardingConfiguration
+            {
+                Shards = new DatabaseTopology[RestoreConfiguration.ShardRestoreSettings.Length],
+                Orchestrator = new OrchestratorConfiguration
+                {
+                    Topology = new OrchestratorTopology
+                    {
+                        Members = new List<string>()
+                    }
+                }
+            };
+
+            var nodes = new HashSet<string>();
+            for (var i = 0; i < RestoreConfiguration.ShardRestoreSettings.Length; i++)
+            {
+                var shardRestoreSetting = RestoreConfiguration.ShardRestoreSettings[i];
+                databaseRecord.Sharding.Shards[i] = new DatabaseTopology
+                {
+                    Members = new List<string>
+                    {
+                        shardRestoreSetting.NodeTag
+                    }
+                };
+
+                if (nodes.Add(shardRestoreSetting.NodeTag))
+                    databaseRecord.Sharding.Orchestrator.Topology.Members.Add(shardRestoreSetting.NodeTag);
+            }
+        }
+
+        private async Task<IOperationResult> RestoreOnAllShards(Action<IOperationProgress> onProgress, RestoreBackupConfigurationBase configuration,
+            ShardedDatabaseContext shardedDatabaseContext)
+        {
+            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Aviv, DevelopmentHelper.Severity.Normal, "try to replace this with a proper impl. of ServerStore.AddRemoteOperation");
```

Comment:
want me to assist you with that?

Reply:
sure thanks, but let's handle that in a separate PR : 
https://issues.hibernatingrhinos.com/issue/RavenDB-19020/Sharding-Restore-Replace-RestoreOnAllShards-with-an-Implementation-of-AddRemoteOperation-for-ServerWide-operations